### DISCUSSION TITLE
Is it possible to compile to web target?

### DISCUSSION THREAD

**Discussion by anissen** [foreignsasquatch/raylib-hx#6] [Q&A] [answered]
Raylib can be compiled to JavaScript via Emscripten (https://github.com/raysan5/raylib/wiki/Working-for-Web-(HTML5)). Is it possible to somehow get a JavaScript-target version of client code + raylib using this library?

**Comment by foreignsasquatch** [comment]
I don't think its possible since raylib requires a special build of it using emcc but it's worth a try. I will try to work on it when I have time 🙂

**Reply by anissen** [reply]
Thanks for the reply! I was thinking that it might be possible by also bundling a JavaScript build of Raylib. The Haxe extern definitions would still be the same but would (in theory) make it possible to build the Haxe code to the JavaScript target and run Raylib-powered code on a website by embedding both the Raylib script and the Haxe JavaScript output. It's an idea, at least 😅 

I'm asking because Raylib has a _very_ nice, ergonomic, API to program against but being restricted to the C++ target is too bit a limiting for many, myself included.

**Reply by foreignsasquatch** [reply]
I would really like raylib on the Web too but emscripten won't probably work. I'm open to any idea rn to add Web support 😄

**Reply by l0go** [reply]
Emscripten doesn't get compiled to JavaScript, it gets compiled into WebAssembly. You might be able to compile the generated C++ code using Emscripten though?

**Reply by l0go** [reply]
Looks like we might be able to make a custom build.xml file with: 
```<setup name="emscripten" />```

**Reply by anissen** [reply]
My idea was to take a more pragmatic approach and simply bundle e.g. `raylib.wasm` with this library. The Haxe part would then be coded against the existing extern definitions and result in e.g. `myapp.js`.

An `index.html` could then be made where
```html
<html>
...
<script src="raylib.wasm" />
<script src="myapp.js" />
...
```

It may not work at all in practice, but it seems like a simple™️ solution.

**Reply by l0go** [reply]
> My idea was to take a more pragmatic approach and simply bundle e.g. `raylib.wasm` with this library. The Haxe part would then be coded against the existing extern definitions and result in e.g. `myapp.js`.
> 
> An `index.html` could then be made where
> 
> ```
> <html>
> ...
> <script src="raylib.wasm" />
> <script src="myapp.js" />
> ...
> ```
> 
> It may not work at all in practice, but it seems like a simple™️ solution.

Issue with that, is that it would require a binary blob in the project which isn't something we would want. Preferably it should be compiled using Haxe.

**Comment by foreignsasquatch** [comment]
@anissen Work being done on hxRaylib for web in the experimental branch :)

**Reply by foreignsasquatch** [reply]
thanks to @rc-05

**Reply by ghost** [reply]
For now I am in the processing of porting rlgl.h to Haxe (which is a pseudo OpenGL 1.1 API) that raylib uses internally in order to simplify porting for the web
I'm working on it on-off for now tho @anissen

**Comment by foreignsasquatch** [comment, marked as answer]
WASM+hxRaylib :)
![image](https://user-images.githubusercontent.com/74960320/148897330-f23a8eec-6a8b-458c-ab40-a1797a22f206.png)
will push soon

**Reply by foreignsasquatch** [reply]
with a few flags you can achieve this :D
```hxml
-D wasm
-D EMSCRIPTEN_SDK="C:/Users/DELL/Desktop/Env/emsdk/upstream/emscripten"
-D HXCPP_LINK_EMSCRIPTEN_EXT=.html
-D emscripten
````

**Reply by foreignsasquatch** [reply]
you can use you're own html file instead of the generated one, btw

**Reply by foreignsasquatch** [reply]
i suggest you dont use it for now since its a bit hacky right now

**Reply by anissen** [reply]
It works! 😲 🎉 

![image](https://user-images.githubusercontent.com/879967/148912835-10de591c-98b9-404e-b9b5-84fec0e55849.png)

On macOS at least, the `EMSCRIPTEN_SDK` define must be without quotes for it to work.

For instance `-D EMSCRIPTEN_SDK="/Users/anders/projects/misc/emsdk/upstream/emscripten"` results in 
```
sh: emcc: command not found
Error: Could not create PCH
```

Using `EMSCRIPTEN_SDK` without quotes makes `emcc` be found.

**Reply by foreignsasquatch** [reply]
oh nice! images and other assets wont work for now on the latest commit 😓 , i have it fixed! More commits coming soon:tm:!

**Reply by foreignsasquatch** [reply]
Now you can do `-D ASSET_PATH=path/to/assets@assets` and load it by `LoadImage(assets/something.png);`!

**Reply by ghost** [reply]
This is great news! It's great to see hxRaylib running on the web with emscripten :smiley:
